consensus/istanbul/backend/engine.go

```diff
@@ -560,10 +561,53 @@ func (sb *Backend) SetChain(chain consensus.ChainReader, currentBlock func() *ty
 	sb.chain = chain
 	sb.currentBlock = currentBlock
 	sb.stateAt = stateAt
+
+	if bc, ok := chain.(*ethCore.BlockChain); ok {
+		go sb.newChainHeadLoop(bc)
+		go sb.updateReplicaStateLoop(bc)
+	}
+
 }
 
-// StartValidating implements consensus.Istanbul.StartValidating
-func (sb *Backend) StartValidating(hasBadBlock func(common.Hash) bool,
+func (sb *Backend) newChainHeadLoop(bc *ethCore.BlockChain) {
+	// Batched. For stats & announce
+	chainHeadCh := make(chan ethCore.ChainHeadEvent, 10)
+	chainHeadSub := bc.SubscribeChainHeadEvent(chainHeadCh)
+	defer chainHeadSub.Unsubscribe()
+
+	for {
+		select {
+		case chainHeadEvent := <-chainHeadCh:
+			sb.newChainHead(chainHeadEvent.Block)
+		case err := <-chainHeadSub.Err():
+			log.Error("Error in istanbul's subscription to the blockchain's chainhead event", "err", err)
+			return
+		}
+	}
+}
+
+func (sb *Backend) updateReplicaStateLoop(bc *ethCore.BlockChain) {
+	// Unbatched event listener
+	chainEventCh := make(chan ethCore.ChainEvent, 10)
+	chainEventSub := bc.SubscribeChainEvent(chainEventCh)
+	defer chainEventSub.Unsubscribe()
+
+	for {
+		select {
+		case chainEvent := <-chainEventCh:
+			if !sb.coreStarted {
+				n := new(big.Int).Add(chainEvent.Block.Number(), common.Big1)
+				sb.replicaState.NewChainHead(n)
+			}
+		case err := <-chainEventSub.Err():
+			log.Error("Error in istanbul's subscription to the blockchain's chain event", "err", err)
```

Comment:
should we panic here?

it seems that after this happens replican won't ever become primary

Reply:
One thing about a panic here, is that I think that this is the normal shut down path, though it might make sense to both add a close channel and understand what occurs with this error.

I don't think I've ran into this causing issues, so I'm less worried about this.